File: dsa-practice-ram/dsa-practice-ram/Arrays/SubArraySum.cs
namespace dsa_practice_ram.Arrays
{
    /*
     *
    - Common Approaches for "Subarrays with Sum K" Problems
    - Prefix Sum + HashMap (O(N)) ->
          Best for handling negative numbers & unordered elements
          This method efficiently tracks cumulative sums and finds subarrays in linear time.
          Works universally even with negatives or when numbers appear in a non-contiguous manner.

    - Sliding Window (O(N) ->
        Only for Positive Numbers) -> Works only for positive numbers
        The sum always increases, so adjusting the window optimally is possible.
        Fails with negatives because shrinking the window may miss valid subarrays.

    - Brute Force (O(N^2)) -> Simple but slow
        Generates all subarrays and checks sums.
        Not efficient for large arrays.

    - Kadane's Algorithm (For Largest Sum Subarray) -> Not directly relevant here but useful for maximum sum problems
    - Helps find the largest sum contiguous subarray efficiently

     * */
    internal class SubArraySum
    {
        public void FindLongestSubArraySum()
        {
            // Find the longest subarray whose sum is equal to k
            // Sliding window works only for positive numbers. For negative numbers use prefix sum
            int[] array = { 1, 2, 3, 0, 0, 0, 1, 1, 1, 1, 4, 2, 3 };
            const int k = 3;
            int sum = 0;
            int windowStart = 0;
            (int start, int end) subArray = default;

            for (int windowEnd = 0; windowEnd < array.Length; windowEnd++)
            {
                sum += array[windowEnd];

                while (sum > k)
                {
                    sum -= array[windowStart];
                    windowStart++;
                }

                if (sum == k)
                {
                    int currentLength = windowEnd - windowStart + 1;

                    if (currentLength > subArray.end - subArray.start + 1)
                        subArray = (windowStart, windowEnd);
                }
            }

            Console.WriteLine(subArray);
        }

        public void CountSubArraysWithSum()
        {
            // Count the number of subarrays whose sum equal to k
            // Sliding window works only for positive numbers. For negagive numbers use prefix sum
            int[] array = { 1, 2, 3, 3, 1, 1, 1, 4, 3, 2, 3 };

            const int k = 3;

            int windowStart = 0;
            int windowSum = 0;
            int count = 0;

            for (int windowEnd = 0; windowEnd < array.Length; windowEnd++)
            {
                windowSum += array[windowEnd];

                while (windowSum > k)
                {
                    windowSum -= array[windowStart];
                    windowStart++;
                }

                if (windowSum == k)
                {
                    count++;
                    Console.WriteLine($"{windowStart}, {windowEnd}");
                }
            }
            Console.WriteLine(count);
        }

        public void CountSubArraysWithNegativeNumbers()
        {
            // Count the number of subarrays whose sum equal to k
            // Prefix sum method

            int[] array = { 1, 2, 0, 0, 3, -3, 1, 1, 1, 4, 3, 2, -3 };

            const int k = 3;

            Dictionary<int, int> prefixSumCount = new();

            // If a prefix sum ever becomes exactly k, then the subarray from index 0 to current index should be counted."
            // To account the  beginning of the array, when prefixsum == k
            prefixSumCount[0] = 1;
            int prefixSum = 0;
            int count = 0;

            for (int i = 0; i < array.Length; i++)
            {
                prefixSum += array[i];

                int target = prefixSum - k;

                if (prefixSumCount.ContainsKey(target))
                {
                    count += prefixSumCount[target];
                }

                if (!prefixSumCount.ContainsKey(prefixSum))
                    prefixSumCount[prefixSum] = 0;

                prefixSumCount[prefixSum]++;
            }

            Console.WriteLine(count);
        }

        public void CalculateSumBetweenTwoIndices()
        {
            int[] nums = { 10, 20, 5, 15 };

            int[] prefixSum = new int[nums.Length];

            prefixSum[0] = nums[0];

            for (int i = 1; i < nums.Length; i++)
            {
                prefixSum[i] = prefixSum[i - 1] + nums[i];
            }

            int start = 0;
            int end = 2;
            int sum = 0;

            if (start == 0)
                sum = prefixSum[end];
            else
                sum = prefixSum[end] - prefixSum[start - 1];

            Console.WriteLine($"Sum between index 0 and 1 is: {prefixSum[2] - prefixSum[0]}");
        }

        public void PrintSubArraysWhichSumK()
        {
            // Print all the subarrays whose sum equal to k
            // Prefix sum method

            int[] array = { 1, 2, 0, 0, 3, -3, 1, 1, 1, 4, 3, 2, -3 };

            const int k = 3;

            Dictionary<int, List<int>> prefixSumMap = new();

            // If a prefix sum ever becomes exactly k, then the subarray from index 0 to current index should be counted."
            // To account the  beginning of the array, when prefixsum == k
            prefixSumMap[0] = new List<int> { -1 };
            int prefixSum = 0;
            List<Range> subArrays = new();

            for (int i = 0; i < array.Length; i++)
            {
                prefixSum += array[i];

                int target = prefixSum - k;

                if (prefixSumMap.ContainsKey(target))
                {
                    foreach (var start in prefixSumMap[target])
                    {
                        subArrays.Add((start + 1)..(i + 1));
                    }
                }

                if (!prefixSumMap.ContainsKey(prefixSum))
                {
                    prefixSumMap[prefixSum] = new List<int>();
                }
                prefixSumMap[prefixSum].Add(i);
            }

            foreach (var range in subArrays)
            {
                Console.WriteLine(string.Join(", ", array[range]));
            }
        }
    }
}

File: MyLearning/PythonPractice/dsa_practice/sliding_window/sub_array_sum.py
"""Subarray sum utilities."""

# C# SubArraySum.FindLongestSubArraySum

def find_longest_subarray_sum(array=None, k=3):
    """Return the longest subarray indices whose sum equals k.

    Problem: Find the longest subarray with sum k (positive numbers).
    Approach: Sliding window with shrinking when the sum exceeds k.
    Time complexity: O(n)
    Space complexity: O(1)
    """
    if array is None:
        array = [1, 2, 3, 0, 0, 0, 1, 1, 1, 1, 4, 2, 3]

    window_start = 0
    current_sum = 0
    best = (0, -1)

    for window_end, value in enumerate(array):
        current_sum += value

        while current_sum > k and window_start <= window_end:
            current_sum -= array[window_start]
            window_start += 1

        if current_sum == k:
            current_length = window_end - window_start + 1
            best_length = best[1] - best[0] + 1
            if current_length > best_length:
                best = (window_start, window_end)

    return best


# C# SubArraySum.CountSubArraysWithSum

def count_subarrays_with_sum(array=None, k=3):
    """Return the count and index ranges for subarrays whose sum equals k.

    Problem: Count subarrays with sum k (positive numbers).
    Approach: Sliding window; shrink when the sum exceeds k.
    Time complexity: O(n)
    Space complexity: O(1)
    """
    if array is None:
        array = [1, 2, 3, 3, 1, 1, 1, 4, 3, 2, 3]

    window_start = 0
    window_sum = 0
    ranges = []

    for window_end, value in enumerate(array):
        window_sum += value

        while window_sum > k and window_start <= window_end:
            window_sum -= array[window_start]
            window_start += 1

        if window_sum == k:
            ranges.append((window_start, window_end))

    return len(ranges), ranges


# C# SubArraySum.CountSubArraysWithNegativeNumbers

def count_subarrays_with_negative_numbers(array=None, k=3):
    """Return the count of subarrays whose sum equals k.

    Problem: Count subarrays with sum k (supports negatives).
    Approach: Prefix sum with a dictionary of counts.
    Time complexity: O(n)
    Space complexity: O(n)
    """
    if array is None:
        array = [1, 2, 0, 0, 3, -3, 1, 1, 1, 4, 3, 2, -3]

    prefix_sum_count = {0: 1}
    prefix_sum = 0
    count = 0

    for value in array:
        prefix_sum += value
        target = prefix_sum - k

        count += prefix_sum_count.get(target, 0)
        prefix_sum_count[prefix_sum] = prefix_sum_count.get(prefix_sum, 0) + 1

    return count


# C# SubArraySum.CalculateSumBetweenTwoIndices

def calculate_sum_between_two_indices(nums=None, start=0, end=2):
    """Return the sum between two indices using prefix sums."""
    if nums is None:
        nums = [10, 20, 5, 15]

    if not nums or start < 0 or end >= len(nums) or start > end:
        return 0

    prefix_sum = [0] * len(nums)
    prefix_sum[0] = nums[0]

    for i in range(1, len(nums)):
        prefix_sum[i] = prefix_sum[i - 1] + nums[i]

    if start == 0:
        return prefix_sum[end]
    return prefix_sum[end] - prefix_sum[start - 1]


# C# SubArraySum.PrintSubArraysWhichSumK

def subarrays_which_sum_k(array=None, k=3):
    """Return all subarrays whose sum equals k (supports negatives).

    Problem: Find all subarrays with sum k.
    Approach: Prefix sum map that tracks all previous indices.
    Time complexity: O(n + m) where m is number of results
    Space complexity: O(n)
    """
    if array is None:
        array = [1, 2, 0, 0, 3, -3, 1, 1, 1, 4, 3, 2, -3]

    prefix_sum_map = {0: [-1]}
    prefix_sum = 0
    subarrays = []

    for i, value in enumerate(array):
        prefix_sum += value
        target = prefix_sum - k

        if target in prefix_sum_map:
            for start in prefix_sum_map[target]:
                subarrays.append(list(array[start + 1:i + 1]))

        prefix_sum_map.setdefault(prefix_sum, []).append(i)

    return subarrays
